SSWC/doc/man/man3/Frame.3
.TH "Frame" 3 "Mon May 9 2016" "Version 0.1" "MissionsVisualizer" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Frame \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <Frame\&.h>\fP
.SS "Classes"

.in +1c
.ti -1c
.RI "struct \fBFrameInfo\fP"
.br
.in -1c
.SS "Public Types"

.in +1c
.ti -1c
.RI "enum \fBFrameType\fP { \fBFT_INERTIAL\fP = 1, \fBFT_PCK\fP, \fBFT_CK\fP, \fBFT_TK\fP, \fBFT_DYNAMIC\fP }"
.br
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBFrame\fP (\fBint\fP spiceId, const \fBstd::string\fP &\fBname\fP='')"
.br
.ti -1c
.RI "\fBFrame\fP (const \fBstd::string\fP &spiceName, const \fBstd::string\fP &\fBname\fP='')"
.br
.ti -1c
.RI "\fB~Frame\fP ()"
.br
.ti -1c
.RI "long \fBGetSpiceId\fP () const "
.br
.ti -1c
.RI "\fBstd::string\fP \fBGetSpiceName\fP () const "
.br
.ti -1c
.RI "const \fBstd::string\fP & \fBGetName\fP () const "
.br
.ti -1c
.RI "\fBFrameInfo\fP \fBGetFrameInfo\fP () const "
.br
.ti -1c
.RI "const \fBSpaceObject\fP & \fBGetCenterObject\fP () const "
.br
.ti -1c
.RI "\fBVector3\fP \fBTransformVector\fP (const \fBVector3\fP &vec, const \fBDate\fP &\fBt\fP, const \fBFrame\fP &\fBref\fP) const "
.br
.ti -1c
.RI "\fBVector3\fP \fBAxisX\fP (const \fBDate\fP &\fBt\fP, const \fBFrame\fP &\fBref\fP) const "
.br
.ti -1c
.RI "\fBVector3\fP \fBAxisY\fP (const \fBDate\fP &\fBt\fP, const \fBFrame\fP &\fBref\fP) const "
.br
.ti -1c
.RI "\fBVector3\fP \fBAxisZ\fP (const \fBDate\fP &\fBt\fP, const \fBFrame\fP &\fBref\fP) const "
.br
.ti -1c
.RI "\fBMatrix4x4\fP \fBGetTransformationMatrix\fP (const \fBDate\fP &\fBt\fP, const \fBFrame\fP &\fBref\fP) const "
.br
.ti -1c
.RI "bool \fBHasAvailableData\fP () const "
.br
.ti -1c
.RI "bool \fBHasLimitedCoverage\fP () const "
.br
.ti -1c
.RI "\fBWindow\fP \fBGetCoverage\fP () const "
.br
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static bool \fBValidateId\fP (long \fBid\fP)"
.br
.ti -1c
.RI "static long \fBMakeFrameId\fP (\fBFrameType\fP \fBtype\fP, long classId)"
.br
.ti -1c
.RI "static std::vector< long > \fBGetBuiltInIds\fP ()"
.br
.ti -1c
.RI "static std::vector< long > \fBGetPoolIds\fP ()"
.br
.ti -1c
.RI "static std::vector< long > \fBGetLoadedPckIds\fP ()"
.br
.in -1c
.SS "Static Public Attributes"

.in +1c
.ti -1c
.RI "static const \fBFrame\fP \fBJ2000\fP = \fBFrame\fP('J2000')"
.br
.ti -1c
.RI "static const \fBFrame\fP \fBECLIPJ2000\fP = \fBFrame\fP('ECLIPJ2000')"
.br
.in -1c
.SH "Detailed Description"
.PP 
Definition at line 14 of file Frame\&.h\&.
.SH "Member Enumeration Documentation"
.PP 
.SS "enum \fBFrame::FrameType\fP"

.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fIFT_INERTIAL \fP\fP
.TP
\fB\fIFT_PCK \fP\fP
.TP
\fB\fIFT_CK \fP\fP
.TP
\fB\fIFT_TK \fP\fP
.TP
\fB\fIFT_DYNAMIC \fP\fP
.PP
Definition at line 17 of file Frame\&.h\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "Frame::Frame (\fBint\fPspiceId, const \fBstd::string\fP &name = \fC''\fP)"

.PP
Definition at line 6 of file Frame\&.cpp\&.
.SS "Frame::Frame (const \fBstd::string\fP &spiceName, const \fBstd::string\fP &name = \fC''\fP)"

.PP
Definition at line 11 of file Frame\&.cpp\&.
.SS "Frame::~Frame ()"

.PP
Definition at line 34 of file Frame\&.cpp\&.
.SH "Member Function Documentation"
.PP 
.SS "\fBVector3\fP Frame::AxisX (const \fBDate\fP &t, const \fBFrame\fP &ref) const"

.PP
Definition at line 93 of file Frame\&.cpp\&.
.SS "\fBVector3\fP Frame::AxisY (const \fBDate\fP &t, const \fBFrame\fP &ref) const"

.PP
Definition at line 98 of file Frame\&.cpp\&.
.SS "\fBVector3\fP Frame::AxisZ (const \fBDate\fP &t, const \fBFrame\fP &ref) const"

.PP
Definition at line 103 of file Frame\&.cpp\&.
.SS "std::vector< long > Frame::GetBuiltInIds ()\fC [static]\fP"

.PP
Definition at line 190 of file Frame\&.cpp\&.
.SS "const \fBSpaceObject\fP & Frame::GetCenterObject () const"

.PP
Definition at line 74 of file Frame\&.cpp\&.
.SS "\fBWindow\fP Frame::GetCoverage () const"

.PP
Definition at line 151 of file Frame\&.cpp\&.
.SS "\fBFrame::FrameInfo\fP Frame::GetFrameInfo () const"

.PP
Definition at line 57 of file Frame\&.cpp\&.
.SS "std::vector< long > Frame::GetLoadedPckIds ()\fC [static]\fP"

.PP
Definition at line 204 of file Frame\&.cpp\&.
.SS "const \fBstd::string\fP & Frame::GetName () const"

.PP
Definition at line 52 of file Frame\&.cpp\&.
.SS "std::vector< long > Frame::GetPoolIds ()\fC [static]\fP"

.PP
Definition at line 197 of file Frame\&.cpp\&.
.SS "long Frame::GetSpiceId () const"

.PP
Definition at line 39 of file Frame\&.cpp\&.
.SS "\fBstd::string\fP Frame::GetSpiceName () const"

.PP
Definition at line 44 of file Frame\&.cpp\&.
.SS "\fBMatrix4x4\fP Frame::GetTransformationMatrix (const \fBDate\fP &t, const \fBFrame\fP &ref) const"

.PP
Definition at line 108 of file Frame\&.cpp\&.
.SS "bool Frame::HasAvailableData () const"

.PP
Definition at line 128 of file Frame\&.cpp\&.
.SS "bool Frame::HasLimitedCoverage () const"

.PP
Definition at line 142 of file Frame\&.cpp\&.
.SS "long Frame::MakeFrameId (\fBFrameType\fPtype, longclassId)\fC [static]\fP"

.PP
Definition at line 179 of file Frame\&.cpp\&.
.SS "\fBVector3\fP Frame::TransformVector (const \fBVector3\fP &vec, const \fBDate\fP &t, const \fBFrame\fP &ref) const"

.PP
Definition at line 79 of file Frame\&.cpp\&.
.SS "bool Frame::ValidateId (longid)\fC [static]\fP"

.PP
Definition at line 167 of file Frame\&.cpp\&.
.SH "Member Data Documentation"
.PP 
.SS "const \fBFrame\fP Frame::ECLIPJ2000 = \fBFrame\fP('ECLIPJ2000')\fC [static]\fP"

.PP
Definition at line 73 of file Frame\&.h\&.
.SS "const \fBFrame\fP Frame::J2000 = \fBFrame\fP('J2000')\fC [static]\fP"

.PP
Definition at line 72 of file Frame\&.h\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for MissionsVisualizer from the source code\&.


SSWC/src/CSpice/Frame.cpp
#include "Frame.h"

const Frame Frame::J2000 = Frame("J2000");
const Frame Frame::ECLIPJ2000 = Frame("ECLIPJ2000");

Frame::Frame(int spiceId, const std::string& name)
{
	Construct(spiceId, name);
}

Frame::Frame(const std::string& spiceName, const std::string& name)
{
	long spiceId;
	CSPICE_ASSERT( namfrm_c(spiceName.c_str(), &spiceId) );

	Construct(spiceId, name);
}

void Frame::Construct(int spiceId, const std::string& name)
{
	if(!ValidateId(spiceId))
		CSpiceUtil::SignalError("No such CSpice frame is defined");

	this->spiceId = spiceId;

	if(name != "")
		this->name = name;
	else
		this->name = GetSpiceName();

	this->centerObject = new SpaceObject(GetFrameInfo().centerId, this->name + "_center");
}

Frame::~Frame()
{
	
}

long Frame::GetSpiceId() const
{
	return spiceId;
}

std::string Frame::GetSpiceName() const
{
	char frameName[FRAME_NAME_MAX_LENGTH];
	CSPICE_ASSERT( frmnam_c(spiceId, FRAME_NAME_MAX_LENGTH, frameName) );

	return std::string(frameName);
}

const std::string& Frame::GetName() const
{
	return name;
}

Frame::FrameInfo Frame::GetFrameInfo() const
{
	SpiceInt centerId;
	SpiceInt clssid;
	SpiceInt frclss;
	SpiceBoolean found;

	CSPICE_ASSERT(frinfo_c(spiceId, &centerId, &frclss, &clssid, &found));

	FrameInfo finfo;
	finfo.centerId = centerId;
	finfo.classId = clssid;
	finfo.frameType = FrameType(frclss);

	return finfo;
}

const SpaceObject& Frame::GetCenterObject() const
{
	return *centerObject;
}

Vector3 Frame::TransformVector(const Vector3& vec, const Date& t, const Frame& ref) const
{
	double transform[3][3];

	CSPICE_ASSERT(pxform_c(GetSpiceName().c_str(), ref.GetSpiceName().c_str(), t.AsDouble(), transform));

	double axisLocal[3] = {vec.x, vec.y, vec.z};
	double axisGlobal[3];

	CSPICE_ASSERT(mxv_c(transform, axisLocal, axisGlobal));

	return Vector3(axisGlobal);
}

Vector3 Frame::AxisX(const Date& t, const Frame& ref) const
{
	return TransformVector(Vector3::i, t, ref);
}

Vector3 Frame::AxisY(const Date& t, const Frame& ref) const
{
	return TransformVector(Vector3::j, t, ref);
}

Vector3 Frame::AxisZ(const Date& t, const Frame& ref) const
{
	return TransformVector(Vector3::k, t, ref);
}

Matrix4x4 Frame::GetTransformationMatrix(const Date& t, const Frame& ref) const
{
	Matrix4x4 transform = Matrix4x4::Zero;

	Vector3 x = AxisX(t, ref);
	Vector3 y = AxisY(t, ref);
	Vector3 z = AxisZ(t, ref);

	for(int row = 0; row < 3; row++)
	{
		transform.Set(row, 0, x[row]);
		transform.Set(row, 1, y[row]);
		transform.Set(row, 2, z[row]);
	}

	transform.Set(3, 3, 1.0);

	return transform;
}

bool Frame::HasAvailableData() const
{
	long centerId = GetFrameInfo().centerId;

	bool hasPoleRa = (bodfnd_c(centerId, "POLE_RA") != SPICEFALSE);
	bool hasPoleDec = (bodfnd_c(centerId, "POLE_DEC") != SPICEFALSE);
	bool hasPm = (bodfnd_c(centerId, "PM") != SPICEFALSE);
	bool centerHasIAUFrameData = hasPoleRa && hasPoleDec && hasPm;

	bool isIAUFrame = GetSpiceName().substr(0, 3) == "IAU";

	return (isIAUFrame && centerHasIAUFrameData) || HasLimitedCoverage();
}

bool Frame::HasLimitedCoverage() const
{
	const std::vector<long>& pckIds = GetLoadedPckIds();

	std::vector<long>::const_iterator it = std::find(pckIds.begin(), pckIds.end(), this->spiceId);

	return it != pckIds.end();
}

Window Frame::GetCoverage() const
{
	std::vector<KernelData> kernels = CSpiceUtil::GetLoadedKernels("PCK");

	Window coverage;

	for(size_t i = 0; i < kernels.size(); i++)
	{
		std::string file = kernels[i].filename;

		CSPICE_ASSERT(pckcov_c(file.c_str(), this->GetFrameInfo().classId, &coverage.GetSpiceCell()));
	}

	return coverage;
}

bool Frame::ValidateId(long id)
{
	SpiceInt centerId;
	SpiceInt clssid;
	SpiceInt frclss;
	SpiceBoolean found;

	CSPICE_ASSERT(frinfo_c(id, &centerId, &frclss, &clssid, &found));

	return found != SPICEFALSE;
}

long Frame::MakeFrameId(FrameType type, long classId)
{
	long frameId;
	char frameName[FRAME_NAME_MAX_LENGTH];
	long centerId;
	SpiceBoolean found;
	CSPICE_ASSERT(ccifrm_c(type, classId, FRAME_NAME_MAX_LENGTH, &frameId, frameName, &centerId, &found));

	return (found != SPICEFALSE) ? frameId : 0;
}

std::vector<long> Frame::GetBuiltInIds()
{
	SPICEINT_CELL(cell, CELL_SIZE_LARGE);
	CSPICE_ASSERT(bltfrm_c(SPICE_FRMTYP_ALL, &cell));
	return CSpiceUtil::IntCellToVector(cell);
}

std::vector<long> Frame::GetPoolIds()
{
	SPICEINT_CELL(cell, CELL_SIZE_LARGE);
	CSPICE_ASSERT(kplfrm_c(SPICE_FRMTYP_ALL, &cell));
	return CSpiceUtil::IntCellToVector(cell);
}

std::vector<long> Frame::GetLoadedPckIds()
{
	SPICEINT_CELL(cell, CELL_SIZE_LARGE);
	std::vector<KernelData> kernels = CSpiceUtil::GetLoadedKernels("PCK");
	for(size_t i = 0; i < kernels.size(); i++)
	{
		CSPICE_ASSERT(pckfrm_c(kernels[i].filename.c_str(), &cell));
	}

	std::vector<long> classIds = CSpiceUtil::IntCellToVector(cell);
	std::vector<long> frameIds;
	frameIds.reserve(classIds.size());

	for(size_t i = 0; i < classIds.size(); i++)
	{
		frameIds.push_back(Frame::MakeFrameId(FT_PCK, classIds[i]));
	}

	return frameIds;
	
}
